libvips/foreign/foreign.c

```diff
@@ -1669,34 +1705,80 @@ vips__foreign_convert_saveable(VipsImage *in, VipsImage **ready,
 		in = out;
 	}
 
+	*ready = in;
+
+	return 0;
+}
+
+/* Map VipsForeignPreserve flags to metadata names.
+ */
+typedef struct _VipsForeignPreserveNames {
+	VipsForeignPreserve flag;
+	const char *name;
+} VipsForeignPreserveNames;
+
+const VipsForeignPreserveNames vips__preserve_names[] = {
+	{ VIPS_FOREIGN_PRESERVE_EXIF, VIPS_META_EXIF_NAME },
+	{ VIPS_FOREIGN_PRESERVE_XMP, VIPS_META_XMP_NAME },
+	{ VIPS_FOREIGN_PRESERVE_IPTC, VIPS_META_IPTC_NAME },
+	{ VIPS_FOREIGN_PRESERVE_ICC, VIPS_META_ICC_NAME }
+};
+
+static void *
+vips_foreign_save_remove_other(VipsImage *image,
+	const char *field, GValue *value, void *user_data)
+{
+	if (vips_isprefix("png-comment-", field) ||
+		strcmp(field, VIPS_META_PHOTOSHOP_NAME) == 0 ||
+		strcmp(field, VIPS_META_IMAGEDESCRIPTION) == 0)
```

Comment:
This looks hard to maintain :( 

How about doing this in a single pass over the metadata? 

At the moment, you have one loop over the array of preserve names, then a second loop over all the metadata items for "other" handling.

Instead, only have the `vips_image_map()` loop and in the handler test each field name against preserve names. You could remove these special cases.

Reply:
You're right, fixed with commit e83194b26a86365f0bfa4cbf38baad89d953fcad.